### PR TITLE
SAK-42257 - Problems archiving message forums

### PR DESCRIPTION
**Pull request by adrianmticarum**
- New code to export/import messages from forums (until this PR just forums and discussions were exported/imported).
- Existing code adapted to JPA EntityManagerFactory.

### CODE REVIEW

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/DiscussionForumServiceImpl.java [hunk 1]
```diff
@@ -602,362 +665,444 @@ public String getLabel()
...
+				.map(element -> (Element) element).collect(Collectors.toList());
+	}
+
+	private List<Element> getAttachmentElementList(List<Element> elements) {
+		return elements.stream().filter(e -> ATTACHMENT.equals(e.getTagName()))
+				.collect(Collectors.toList());
+	}
+
+	private void mergeDiscussionTopicPropertiesNodes(final DiscussionTopic discussionTopic,
+													 final Element propertiesElement) {
```
Comment (ern) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/DiscussionForumServiceImpl.java]: spacing

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/DiscussionForumServiceImpl.java [hunk 2]
```diff
@@ -602,362 +665,444 @@ public String getLabel()
...
+		Boolean booleanToReturn = null;
+		if (StringUtils.isNotBlank(stringToConvert)) {
+			booleanToReturn = Boolean.valueOf(stringToConvert);
+		}
+		return booleanToReturn;
+	}
+
+	private Integer nullSafeToInteger(String stringToConvert) {
+		Integer integerToReturn = null;
+		if (StringUtils.isNotBlank(stringToConvert)) {
```
Comment (ern) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/DiscussionForumServiceImpl.java]: Maybe use StringUtils.isNumeric which makes it a little more safer ;)

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java [hunk 3]
```diff
@@ -884,71 +887,78 @@ public DiscussionTopic createDiscussionForumTopic(DiscussionForum forum) {
...
+
+        //make sure availability is set properly
+        topic.setAvailability(ForumScheduleNotificationCover.makeAvailableHelper(topic.getAvailabilityRestricted(), topic.getOpenDate(), topic.getCloseDate()));
+
+        DiscussionTopic topicReturn = topic;
+        if (isNew) {
+            DiscussionForum discussionForum = (DiscussionForum) getForumByIdWithTopics(topic.getBaseForum().getId());
+            discussionForum.addTopic(topic);
+        } else {
+            topicReturn = getHibernateTemplate().merge(topic);
```
Comment (ern) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java]: Given that these services are already advised by spring tx lets use this pattern:
`getSessionFactory().getCurrentSession().merge(topic);`

Comment (ern) [reply msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java]: eventually we want to remove HibernateTemplate as it's a deprecated pattern.

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java [hunk 4]
```diff
@@ -1312,55 +1322,30 @@ public List getForumByTypeAndContextWithTopicsAllAttachments(final String typeUu
...
-				}
-			}
 
-			List resultList = Util.setToList(resultSet);
-			Collections.sort(resultList, FORUM_SORT_INDEX_CREATED_DATE_COMPARATOR_DESC);
+	public List getForumByTypeAndContextWithTopicsMembership(final String typeUuid, final String contextId) {
+		if (typeUuid == null || contextId == null) {
+			throw new IllegalArgumentException("Null Argument");
+		}
+		final HibernateCallback<List> hcb = session -> {
```
Comment (ern) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java]: `<List>` what? use generics

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsMessageManagerImpl.java [hunk 5]
```diff
@@ -1261,35 +1261,33 @@ public Attachment createAttachment() {
...
         boolean isNew = message.getId() == null;
-        
-        if (!ignoreLockedTopicForum && !(message instanceof PrivateMessage)){                  
-          if (isForumOrTopicLocked(message.getTopic().getBaseForum().getId(), message.getTopic().getId())) {
-              log.info("saveMessage executed [messageId: " + (isNew ? "new" : message.getId().toString()) + "] but forum is locked -- save aborted");
-              throw new LockedException("Message could not be saved [messageId: " + (isNew ? "new" : message.getId().toString()) + "]");
-          }
+
+        if (!ignoreLockedTopicForum && !(message instanceof PrivateMessage) && isForumOrTopicLocked(message.getTopic().getBaseForum().getId(), message.getTopic().getId())) {
+            log.info("saveOrUpdateMessage executed [messageId: " + (isNew ? "new" : message.getId().toString()) + "] but forum is locked -- save aborted");
```
Comment (ern) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsMessageManagerImpl.java]: log message can be improved:
`log.warn("Forum or Topic is locked for [messageId: {}] not saving", (isNew ? "new" : message.getId().toString()));`

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsMessageManagerImpl.java [hunk 6]
```diff
@@ -1303,35 +1301,125 @@ public void saveMessage(Message message, boolean logEvent, String toolId, String
...
+        return saveMessage(message, logEvent, toolManager.getCurrentTool().getId(), getCurrentUser(), getContextId(), ignoreLockedTopicForum);
+    }
+
+    @Override
+    public String saveMessage(Message message, boolean logEvent, String toolId, String userId, String contextId,
+                              boolean ignoreLockedTopicForum) {
+
+        if (!ignoreLockedTopicForum && !(message instanceof PrivateMessage)
+                && isForumOrTopicLocked(message.getTopic().getBaseForum().getId(), message.getTopic().getId())) {
+            log.info("saveMessage executed [messageId: new] but forum is locked -- save aborted");
```
Comment (ern) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsMessageManagerImpl.java]: make this a warn?

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/PermissionLevelManagerImpl.java [hunk 7]
```diff
@@ -302,12 +302,12 @@ public DBMembershipItem createDBMembershipItem(String name, String permissionLev
 		return newDBMembershipItem;		
 	}
   
-  public void saveDBMembershipItem(DBMembershipItem item){
-  	getHibernateTemplate().saveOrUpdate(item);
+  public DBMembershipItem saveDBMembershipItem(DBMembershipItem item){
+		return getHibernateTemplate().merge(item);
```
Comment (ern) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/PermissionLevelManagerImpl.java]: remove the use of HibernateTemplate here and in other places

File: msgcntr/messageforums-api/src/java/org/sakaiproject/api/app/messageforums/MessageForumsMessageManager.java [hunk 8]
```diff
@@ -59,7 +59,34 @@
      * to be updated even if the topic or forum is locked (ie marking as read or
      * commenting on a moderated message)
      */
-    public void saveMessage(Message message, boolean logEvent, boolean ignoreLockedTopicForum);
+    public Message saveOrUpdateMessage(Message message, boolean logEvent, boolean ignoreLockedTopicForum);
+
```
[the remark this answers is not in the document]
Comment (adrianmticarum) [reply msgcntr/messageforums-api/src/java/org/sakaiproject/api/app/messageforums/MessageForumsMessageManager.java]: @return added.

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/DiscussionForumServiceImpl.java (hunk 1 above)
Comment (adrianmticarum) [reply msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/DiscussionForumServiceImpl.java]: Fixed!

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/DiscussionForumServiceImpl.java (hunk 2 above)
Comment (adrianmticarum) [reply msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/DiscussionForumServiceImpl.java]: Yes! Thanks for the tip!

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java (hunk 3 above)
Comment (adrianmticarum) [reply msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java]: Done!

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java (hunk 4 above)
Comment (adrianmticarum) [reply msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsForumManagerImpl.java]: Done!

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsMessageManagerImpl.java (hunk 5 above)
Comment (adrianmticarum) [reply msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsMessageManagerImpl.java]: Right! Done.

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsMessageManagerImpl.java (hunk 6 above)
Comment (adrianmticarum) [reply msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/MessageForumsMessageManagerImpl.java]: Changed!

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/PermissionLevelManagerImpl.java (hunk 7 above)
Comment (adrianmticarum) [reply msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/PermissionLevelManagerImpl.java]: Done where affected!

**Comment by adrianmticarum** [comment]
Hi @ern ! Can you please take a look at this PR? Thanks.

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/DiscussionForumManagerImpl.java [hunk 9]
```diff
@@ -1150,61 +1141,22 @@ public DiscussionForum saveForum(DiscussionForum forum, boolean draft, String co
     }
 
     boolean saveArea = forum.getId() == null;
-    forum.setDraft(Boolean.valueOf(draft));
```
Comment (mpellicer) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/DiscussionForumManagerImpl.java]: THANK YOU!!

**Comment by adrianfish** [comment]
@ern How about it? Good to go?

File: msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java [hunk 10]
```diff
@@ -328,15 +328,15 @@ public PrivateForum initializationHelper(PrivateForum forum, Area area, String u
    */
   public void savePrivateMessage(Message message)
   {
-    messageManager.saveMessage(message);
+    messageManager.saveOrUpdateMessage(message);
```
Comment (ottenhoff) [inline msgcntr/messageforums-component-impl/src/java/org/sakaiproject/component/app/messageforums/ui/PrivateMessageManagerImpl.java]: should we return the Message in this method to be consistent?

**Comment by ottenhoff** [comment]
This looks good to me. I apologize for working on something similar yesterday. I was the likely cause of the conflicts. I should have looked for your changes first.

**Pull request merged**